### PR TITLE
Support toplevel defs with metadata

### LINKED ISSUE #42: Toplevel defs with metadata are not handled properly
This is valid Clojure code and is handled correctly in `clojure-mode`:

```clojure
^{:a 1}
(def b 2)
```

## Expected behavior

To match the behavior of `clojure-mode` which is:
-) `b` can navigated to via imenu
-) `(def b 2)` properly font locked liked other toplevel defs
-) `(def b 2)` not indented

## Actual behavior

-) `b` not included in imenu
-) `(def b 2)` not font locked
-) `(def b 2)` indented 

## Steps to reproduce the problem

Put this a `clojure-ts-mode` buffer:

```clojure
^{:a 1}
(def b 2)
```

### clojure-ts-mode version

```
clojure-ts-mode (version 0.2.2)
```

### tree-sitter-clojure grammar version

`v0.0.12`

### Emacs version

*E.g. 29.3*

### Operating system

macOS 14.4

### PR DESCRIPTION
**Pull request by kommen**
Fixes #42 

Pushing this as WIP and to collect feedback from others if this is the right approach.

- [x] don't indent a toplevel `def` with metadata
- [x] font lock toplevel `def` with metadata
- [x] font lock toplevel `def` symbol name
- [x] make toplevel `def` with metadata findable via imenu
- [x] don't indent a toplevel  vector, set or map with metadata
- [x] font lock docstrings of def/defn forms with metadata

### CODE REVIEW

**Comment by bbatsov** [comment]
I guess that's still waiting for feedback, right?

**Comment by kommen** [comment]
@bbatsov yes, feedback very welcome! however, I'm using this PR in my prod emacs setup and it seems to work as intended. So I would be somewhat confident to merge it for now as is and continue with the remaining todos in a follow up, if no one else has objections.

File: clojure-ts-mode.el [hunk 1]
```diff
@@ -810,6 +810,13 @@ forms like deftype, defrecord, reify, proxy, etc."
            (clojure-ts--match-fn-docstring parent)
            (clojure-ts--match-method-docstring parent))))
 
+(defun clojure-ts--match-toplevel-with-meta (_node parent _bol)
+  "Match NODE when it is toplevel form and it has metadata"
+  (let* ((grandparent (treesit-node-parent parent)))
+    (and (string-equal "source" (treesit-node-type grandparent))
+         (clojure-ts--list-node-p parent)
+         (treesit-node-child-by-field-name parent "meta"))))
```
[the remark this answers is not in the document]
Comment (kommen) [reply clojure-ts-mode.el]: @dannyfreeman thanks. I addressed this and added some test cases.

File: clojure-ts-mode.el [hunk 2]
```diff
@@ -349,7 +349,7 @@ with the markdown_inline grammar."
 
     :feature 'builtin
     :language 'clojure
-    `(((list_lit :anchor (sym_lit (sym_name) @font-lock-keyword-face))
+    `(((list_lit meta: _ :? :anchor (sym_lit (sym_name) @font-lock-keyword-face))
```
[the remark this answers is not in the document]
Comment (kommen) [reply clojure-ts-mode.el:352]: I can't get this to work for now. Shall I revert my change related to this and get the PR merged?

[the remark this answers is not in the document]
Comment (kommen) [reply clojure-ts-mode.el:352]: @dannyfreeman thank you! meanwhile, I had a chance to dive deeper and I think I actually figured it out (mostly?). there are a bunch of similar queries where the the optional metadata node could be added, but I didn't have time to come up with actual clojure code examples testing them so a didn't want to touch them – maybe unnecessarily.

**Comment by kommen** [comment]
Functionality wise this PR now seems complete to me: I think some of the code needs a bit of clarification and better naming.

I will use my recent changes for a while myself in my daily workflow before I will ask for merging this PR. Meanwhile if folks want to give it a try or discuss the changes with me (here or also on the Clojurians slack), I would appreciate this very much.

I'd be also up to split out some changes in their own PR if others are more comfortable to review smaller chunks (ie. font lock fixes, imenu fixes, indentation fixes).

**Comment by bbatsov** [comment]
I agree it'd better if we split this into several smaller PRs.

**Comment by kommen** [comment]
@bbatsov agreed. I will start splitting it into separate PRs once I've had a chance to use it a while all together.

**Comment by kommen** [comment]
As discussed, splitted out the first set of changes to fix the imenu integration here: https://github.com/clojure-emacs/clojure-ts-mode/pull/54

**Comment by kommen** [comment]
And the font locking changes here: https://github.com/clojure-emacs/clojure-ts-mode/pull/55

**Comment by bbatsov** [comment]
Now that #54 and #55 have been merged we can close this PR.

**Comment by kommen** [comment]
There are still the indentation fixes from this PR which I still will separate out and as with #55 try to port some relevant test utilities over from clojure-mode

**Pull request closed without merging**